### DISCUSSION TITLE
Название файлов типа 229_1_file!tmp!_3

### DISCUSSION THREAD

**Discussion by mihail-cornev** [totumonline/totum-ru-issues-and-discussions#534] [Как это сделать?]
У меня есть простая таблица ФАЙЛЫ. Если добавлять файлы прямо из этой простой таблицы, то название файла в системе соответствует оригинальному названию файла, а если добавлять в неё файлы (сразу несколько) через временную таблицу, то названия файлов меняются на вот такие: **...file!tmp!_3**

Подскажите пожалуйста, где может быть проблема?

![Screenshot_11](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/113771444/0ee25cf2-daee-4d80-a2b3-dbb9f234e79d)

**Comment by ttm-alexeypolunin** [comment]
На картинке временная таблица или уже простая?

* Вот здесь описано как переносятся файлы из временной в простую: https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/276

**Reply by mihail-cornev** [reply]
на картинке простая таблица

**Reply by ttm-alexeypolunin** [reply]
Вы делаете set содержимого поле файл из поля во временной в простую. Этого недостаточно для копирования файла, там надо получать его бинарник (здесь где-то была инструкция).

Поэтому у вас остается путь с признаком tmp (который предназначен для того, что бы их потом удалить). А если не удалять, то они последующими файлами перезаписываться будут.

**Reply by mihail-cornev** [reply]
на форуме инструкция где-то? или на сайте тотум в документации?

**Reply by Re-Requeste** [reply]
#276 - пробовали этот метод?

**Reply by ttm-alexeypolunin** [reply]
Вот про нее я и говорил. Спс.

**Reply by mihail-cornev** [reply]
> #276 - пробовали этот метод?

Именно им и воспользовался. Проверю коды, если все сделал как в инструкции.

**Reply by mihail-cornev** [reply]
Насколько я понимаю у меня проблема на первом пункте?
https://youtu.be/D35tqLgm5zI

1. В таблице циклов добавил кнопку, которая открывает окно загрузки файлов:
кнопка "загрузить" (код действия):
`a1= : linkToFileUpload(title: "Загрузить файлы"; limit: 50; code: $code; type: "*"; refresh: true)`
`code:totum
a0=: linkToDataTable(table: 'tmp_upload_files'; title: "Загрузить"; width: 800; height: "80vh"; params: $row; refresh: true)
	row: rowCreate(field: "h_insert" = $#input; field: "ttm_table" = $#nti; field: "ttm_cycle" = $#nci; field: "h_file_from" = #h_file_from)`

2. После загрузки файлов они отправляются в поле insert временной таблицы, и оттуда распределяются по строкам.
поле "h_insert" (код действия):
`= : listReplace(list: $#nfv; action: $val; value: "value")
      val: insert(table: $#ntn; field: 'file' = $filesdata)
        filesdata: listCreate(item: $rowfile)
	  		rowfile: rowCreate(field: "name" = $#value[name]; field: "filestring" = $content)
           		content: fileGetContent(file: $#value[file])`
3. По нажатию кнопки сохранить файлы улетают в текущую таблицу циклов и в простую таблицу:
кнопка "сохранить" (код действия):
`a1= : insertList(table: 'files_'; field: 'file' = $list; field: 'file_ttm_table' = #ttm_table; field: 'file_ttm_cycle' = #ttm_cycle; field: 'file_uploaded_by' = $#nu; field: 'file_from' = #h_file_from;  inserts: "inserts"; log: true)
		list: selectList(table: $#ntn; field: 'file')
a2= : insertList(table: $sel; cycle: #ttm_cycle; field: 'file' = $#inserts; log: true)
		sel: select(table: 'tables'; field: 'name'; where: 'id' = #ttm_table)`

**Reply by Re-Requeste** [reply]
Вот что пишет GPT по Вашему коду: 
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/137899742/46c96192-f91f-4fe6-9d49-2c3d9686c753)


Мне кажется, от того что вы в пункте 2 т.е в h_insert ставите через listReplace нечего не даст, т.к. данные файлы все еще находятся внутри временной таблицы, вам нужно проработать кнопку "сохранить" в пункт 3 и использовать ListReplace при сохранение, а не insertList: 
`
= : ListReplace(list: # file action: $val; value: "value")
//# file = берем просто файлы которые в хедере 
      val: insert(table: 'tmp'; field: 'test11' = $filesdata)
        filesdata: listCreate(item: $rowfile)
	  rowfile: rowCreate(field: "name" = $#value[name]; field: "filestring" = $content)
            content: fileGetContent(file: $#value[file])`

добавьте просто еще один action с другой таблицей с циклов допустим

**Reply by Re-Requeste** [reply]
И у меня вопрос почему так все усложнять, туда сюда по два раза грузить т.е. в обычную где все файлы и внутрь циклов. 
Внутри цикла где файлы просто сделайте h_insert соберите все что вам нужно из таблицы все файлы по фильтрам, по nci и т.п.
Сделайте insert, и сделайте как ссылку str с использованием listJoin + insert[type1][[file]]
А для название используйте f1: SetFormat(text: insert[type1][[name]])

**Reply by mihail-cornev** [reply]
> И у меня вопрос почему так все усложнять, туда сюда по два раза грузить т.е. в обычную где все файлы и внутрь циклов. Внутри цикла где файлы просто сделайте h_insert соберите все что вам нужно из таблицы все файлы по фильтрам, по nci и т.п. Сделайте insert, и сделайте как ссылку str с использованием listJoin + insert[type1][[file]] А для название используйте f1: SetFormat(text: insert[type1][[name]])

Спасибо за совет. Насколько я понимаю, вы предлагаете сделать автозаполнение - https://ru.docs.totum.online/autofill-tables , и в таблице циклов добавить кнопку, по нажатию на которую файлы будут сразу загружаться в простую таблицу. Спасибо, буду пробовать.
про listjoin я не понял, но когда дойду до дела - разберусь

**Reply by mihail-cornev** [reply]
Не пойму как из таблицы циклов загрузить сразу несколько файлов в простую таблицу, такая формула не работает, выдаёт ошибку:

кнопка "загрузить" (код действия):
`a1= : linkToFileUpload(title: "Загрузить файлы"; code: $insert ; limit: 50; type: "*"; refresh: true)`

`insert:totum
= : listReplace(list: $#input; action: $val; value: "value")
      val: insert(table: 'files_'; field: 'file' = $filesdata; field: 'file_ttm_table' = $#nti; field: 'file_ttm_cycle' = $#nci; field: 'file_uploaded_by' = $#nu; log: true)
        filesdata: listCreate(item: $rowfile)
	  		rowfile: rowCreate(field: "name" = $#value[name]; field: "filestring" = $content)
           		content: fileGetContent(file: $#value[file])`

![Screenshot_12](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/113771444/74173501-471f-4593-815e-c992d65ae738)

**Reply by Re-Requeste** [reply]
Вам необходимо вывести определенные строки с файлами в таблицу **расчетная циклов  (ФАЙЛЫ)** , с фильтрациями , и в тоже время чтобы в отдельной **обычной таблице ФАЙЛЫ** хранились данные.

Как бы я это решил. 
1) В таблице  **расчетная циклов ФАЙЛЫ** ,сделал бы h_insert (пример:  SelectRowList) собрал бы id , с учетом where, допустим только определенный пользователь или же только картинки формата .jpg и т.п.
2) В таблице  **расчетная циклов ФАЙЛЫ** , сделал бы кнопку для загрузки в обычную таблицу  выбранные файлы
3) После загрузки, с учетом where в h_insert в данном цикле появились бы эти данные.

---
Для того чтобы вывести h_insert, используйте insert в строчной части. Далее, в type2 где файл можете использовать следующее:
select(table: 'all fiels'; field: 'test11'; where: 'id' = $idn)
idn: str`#insert[id]`
---

С этой логикой, вам не нужно создавать временную таблицу которая загружает вашу систему. Тем самым, мы увидим что в **расчетная циклов  (ФАЙЛЫ)** будут те же файлы которые от фильтровались по тем запросам которые мы задали, и все файлы будут так же хранятся в **отдельной обычной таблице ФАЙЛЫ**.

**Reply by mihail-cornev** [reply]
[см выше](https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/534#discussioncomment-9081047)
У меня проблема именно с кнопкой загрузки нескольких файлов, поэтому я сделал временную таблицу.

**Reply by Re-Requeste** [reply]
Попробуйте используйте скрытое поле.
Код который вы написал не будет работать, т.к. файлы не находятся на сервере (попробовал загрузить разными способами, но понял что они попросту не загружаются на сервер, и из-за этого он не видит их в $#inpute), возможно я ошибаюсь

1. Создайте поле с типом файлы, видно только создателю
2. А на кнопку "ЗАГРУЗИТЬ" повести следующий код: 



//кнопка загрузки и выборки файлов
=:  linkToFileUpload(title: "Загрузить файлы"; code: $insert ; limit: 50; type: "*";)

insert:totum
//загружаем в скрытое поле файлы
a1=: set(table: $ # ntn; field: 'h_testovyy_fayl' = $# input)
//Перебираем список файлов и записываем их в нашу обычную таблицу
a2=: listReplace(list: $selected; action: $val; value: "value")
	selected: select(table: $#ntn; field: 'h_testovyy_fayl') 
//можно обратиться и через @ но мне так удобнее
	val: insert(table: 'type1'; field: 'test11' = $filesdata)
  	filesdata: listCreate(item: $rowfile)
 	rowfile: rowCreate(field: "name" = $#value[name]; field: "filestring" = $content)
   content: fileGetContent(file: $#value[file])
//очищаем наше скрытое поле
 a3=: clear(table: $#ntn; field: 'h_testovyy_fayl')

3. Далее h_insert сам подтянет как и говорил [здесь](https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/534#discussioncomment-9081074)

Это на костылях собрано но должно сработать, конечно же замените все на ваши данные

**Reply by mihail-cornev** [reply]
> 1. Создайте поле с типом файлы, видно только создателю

Спасибо, это сработало. Мои мучения закончились ))
получается система дважды загрузит эти файлы на сервер? или это не так работает?

upd: нет, проблема с Openoffice осталась #524

**Reply by Re-Requeste** [reply]
Да, файлы загружаются 2 раза на сервер. Это на костылях как я и говорил, воспользуйтесь #276 чтобы делать через временные, принцип такой же, как только вы открываете временную таблицу, нажимаете на кнопку ЗАГРУЗИТЬ и данные файла загрузятся в обычный. И через час файлы !tmp удалятся автоматический

Не работал с OpenOffice, не могу знать что к чему

**Reply by mihail-cornev** [reply]
https://youtu.be/Q47hPaLF07Y
Вот такая фигня происходит.

В любом случае спасибо огромное за помощь. Вы очень помогли.

>воспользуйтесь https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/276 чтобы делать через временные, принцип такой же, как только вы открываете временную таблицу, нажимаете на кнопку ЗАГРУЗИТЬ и данные файла загрузятся в обычный.

кстати мы вернёмся в самое начало - с этой проблемой я и пришёл сюда ))

**Reply by Re-Requeste** [reply]
Попробуйте через временную таблицу загрузить таким же кодом, т.е. как у вас было до этого 
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/137899742/d86f413a-ba2c-4c5b-96e2-e55d59d43c41)

Просто поменяйте код в сохранить на примерно [этот](https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/534#discussioncomment-9082487)

**Reply by ttm-alexeypolunin** [reply]
!tmp cделаны для того, что бы они потом удалялись. Это делает функция консольной утилиты bin/totum clean-tmp-tables-files

Если файлы во временных используются в большом количестве, то надо ее поставить в Cron на сервере

1. По ssh заходите на серваек
2. Переключаетесь в totum
```bash
su totum
``` 
3. Открываете Crontab
```bash
crontab -e
```
* Будте очень внимательны тк crontab -r (e и r соседние клавиши) — это удаление всего из крона (плохой дизайн)
* Если первый раз открываете спросит в чем открывать — выбирайте `nano` тк народная мудрость гласит, что из `vi` без 3-х бутылок не выйти потом

4. Добавляете еще одну строку
```bash
*/10 * * * * cd /home/totum/totum-mit/ && bin/totum clean-tmp-tables-files
```
* **Очень важно:** что бы в файле осталась еще одна строка пустая! иначе не будет ваще крон ваще работать (откровенно хуевый дизайн)

5. Сохраняете

`Ctrl` + `X` -> `Y`

Что делает clean-tmp-tables-files:

В предоставленном PHP-коде скрипт удаляет файлы из определенного каталога, основываясь на их имени и времени последнего доступа (атрибуте atime).

Время последнего доступа: используется функция fileatime($fName), которая возвращает время последнего доступа к файлу. Это время сравнивается с текущим временем (time() - 3600), и если файл не был доступен в течение последнего часа, он удаляется.

К сожалению исследование этой функции показало сейчас, что мы не добавили туда обработку каталога защищенных файлов. Поэтому временны + защищенный файл в ней не будут обработаны. Записал на исправление.

При переходе на V5 будем добавлять этот пункт в крон сразу.

**Reply by ttm-alexeypolunin** [reply]
https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/524 там ошибка на стороне тотума.

Мы не тестировали на попытку открыть файл, который имеет путь, отличный от поля, тк там есть механизм (д*я сложный) возврата измененного файла с сервера OO в Тотум.

Записал посмотреть, но исправление нескоро.

**Reply by mihail-cornev** [reply]
> Если файлы во временных используются в большом количестве, то надо ее поставить в Cron на сервере

Сделал как вы написали в инструкции, правда до конца не понял зачем... Получается эти tmp файлы не удалялись с сервера?

> Мы не тестировали на попытку открыть файл, который имеет путь, отличный от поля, тк там есть механизм (д*я сложный) возврата измененного файла с сервера OO в Тотум.

Как я понял мне нет смысла дальше играться с OO, т.е. открывание файлов из таблицы циклов сейчас не получится. А можно как-то имитировать нажатие кнопок "Ред.", "Совместн. ред" через код тотум? То есть добавить кнопки в тотум которые будут запускать действие Ред или Совместн Ред?

![Screenshot_13](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/113771444/7412aad3-fc65-43c2-a09a-258d97b4afcd)


теперь я окончательно запутался во всём что здесь происходит. и не пойму пробовать ли что-то дальше или нет...
